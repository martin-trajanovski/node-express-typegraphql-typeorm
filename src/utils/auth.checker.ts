import * as jwt from 'jsonwebtoken';
import { AuthChecker } from 'type-graphql';

import { Roles } from '@src/graphql/types/roles.enum';
import { Context, UserInterface } from '@src/interfaces';

import { redis } from '.';

const authChecker: AuthChecker<Context, Roles> = async ({ context }, roles) => {
  let authToken = context.req.headers['authorization'];

  if (!authToken) {
    throw new Error('Missing token');
  }

  authToken = authToken.replace('Bearer ', '');

  if (authToken) {
    const secret = process.env.AUTH_TOKEN_SECRET as string;
    try {
      if (redis.getClient && redis.getClient.connected) {
        // NOTE: Check if user is already logged out and the authToken is blacklisted.
        const authTokenBlacklisted = await redis.getAsync(authToken);

        if (authTokenBlacklisted) {
          throw new Error('Wrong credentials');
        }
      }

      const verificationResponse = jwt.verify(
        authToken,
        secret
      ) as UserInterface;
      const user = verificationResponse;

      if (user && user.id) {
        context.user = user;

        if (roles.length === 0) {
          return true;
        }

        if (user.roles.some(role => roles.includes(role))) {
          return true;
        } else {
          throw new Error('Wrong credentials');
        }
      } else {
        throw new Error('Wrong credentials');
      }
    } catch (error) {
      throw new Error('Wrong credentials');
    }
  } else {
    throw new Error('Missing token');
  }
};

export default authChecker;
