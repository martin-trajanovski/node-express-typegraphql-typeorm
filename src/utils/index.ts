export { default as authChecker } from './auth.checker';
export { default as logger } from './logger';
export { default as redis } from './redis';
export { default as validateEnv } from './validateEnv';
