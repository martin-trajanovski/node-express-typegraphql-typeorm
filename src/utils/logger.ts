import { createLogger, format, transports } from 'winston';
const { combine, timestamp, label, colorize, printf, splat } = format;

const myFormat = printf(info => {
  return `${info.timestamp} ${info.label} ${info.level}: ${info.message}`;
});

const logger = createLogger({
  transports: [
    new transports.Console(),
    new transports.File({
      filename: './logs/logs.log',
      level: 'error',
    }),
  ],
  format: combine(
    colorize(),
    label({ label: '[app-server]' }),
    timestamp(),
    splat(),
    myFormat
  ),
});

if (process.env.NODE_ENV !== 'production') {
  logger.debug('Logging initialized at debug level');
}

export default logger;
