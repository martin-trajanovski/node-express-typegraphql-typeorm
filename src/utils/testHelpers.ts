import axios from 'axios';

export const apiCall = async (query: string, token?: string | null) => {
  const config = {
    headers: {},
  };

  if (token) {
    config.headers = {
      Authorization: token,
    };
  }

  // console.log(query);

  // TODO: Maybe later the API url should be formed from env settings but for now it is ok.
  const response = await axios.post(
    'http://localhost:3002/api',
    { query },
    config
  );

  // console.log(response.data);

  return response.data;
};
