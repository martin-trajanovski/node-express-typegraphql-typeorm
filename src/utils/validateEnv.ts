import { cleanEnv, port, str, host } from 'envalid';

function validateEnv(): void {
  cleanEnv(process.env, {
    PORT: port(),
    DB_TYPE: str(),
    DB_PATH: str(),
    DB_PORT: port(),
    DB_PASSWORD: str(),
    DB_USER: str(),
    DB_NAME: str(),
    REDIS_HOST: host(),
    AUTH_TOKEN_SECRET: str(),
    REFRESH_TOKEN_SECRET: str(),
  });
}

export default validateEnv;
