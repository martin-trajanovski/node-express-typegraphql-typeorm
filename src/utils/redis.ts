import { createClient, RedisClient, RedisError } from 'redis';

import logger from './logger';

class Redis {
  private client: RedisClient;

  public create(): void {
    this.client = createClient({
      host: process.env.REDIS_HOST,
    });

    this.client.on('connect', () => {
      logger.info('Redis client connected');
    });

    this.client.on('error', (error: RedisError) => {
      logger.error('Redis not connected: ', error);
    });

    this.client.on('end', () => {
      logger.error('Redis connection has been closed');
    });
  }

  public get getClient(): RedisClient {
    return this.client;
  }

  public getAsync(key: string): Promise<string> {
    return new Promise((resolve, reject) => {
      this.client.get(key, (error, reply) => {
        if (error) {
          logger.error(error);

          reject(error);
        }

        resolve(reply);
      });
    });
  }
}

export default new Redis();
