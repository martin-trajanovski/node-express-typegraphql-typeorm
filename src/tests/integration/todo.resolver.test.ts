import { apiCall } from '@src/utils/testHelpers';

describe('Test todos CRUD', () => {
  let token: string | null = null;
  const query = `
    query {
      todos {
        id
        title
        completed
      }
    }
  `;

  beforeAll(async () => {
    const loginQuery = `
      query {
        login(data: {email: "test@test.com", password: "test123"}) {
          firstName
          token
        }
      }
    `;

    const loginResponse = await apiCall(loginQuery);

    token = loginResponse.data.login.token;
  });

  test('When we try to read the todos, it should get all todos for specific user in deault order', async () => {
    const todosResponse = await apiCall(query, token);
    const { todos } = todosResponse.data;

    expect(todos).toBeDefined();
    expect(todos).toHaveLength(2);
    expect(todos).toMatchSnapshot();
  });

  test('When we try to create a todo, it should return that todo with id', async () => {
    const createTodoMutation = `
      mutation {
        createTodo(data: {title: "New test todo"}) {
          id
          title
          completed
        }
      }
    `;

    const todosResponse = await apiCall(createTodoMutation, token);
    const { createTodo } = todosResponse.data;

    expect(createTodo).toBeDefined();
    expect(createTodo.id).toBeDefined();
    expect(createTodo.title).toBe('New test todo');
    expect(createTodo.completed).toBeFalsy();
  });

  test('When we try to update a todo, it should return that todo updated', async () => {
    const initialTodosResponse = await apiCall(query, token);
    const initialTodos = initialTodosResponse.data.todos;
    const todoToUpdate = initialTodos[initialTodos.length - 1];
    const updateTodoMutation = `
      mutation {
        updateTodo(data: {id: "${todoToUpdate.id}", title: "Test todo updated", completed: true}) {
          id
          title
          completed
        }
      }
    `;

    const removeTodoResponse = await apiCall(updateTodoMutation, token);
    const { updateTodo } = removeTodoResponse.data;

    expect(updateTodo).toBeDefined();
    expect(updateTodo.title).toBe('Test todo updated');
    expect(updateTodo.completed).toBeTruthy();
  });

  test('When we try to remove a todo, it should return that todo', async () => {
    const initialTodosResponse = await apiCall(query, token);
    const initialTodos = initialTodosResponse.data.todos;
    const todoToRemove = initialTodos[initialTodos.length - 1];
    const removeTodoMutation = `
      mutation {
        removeTodo(id: "${todoToRemove.id}")
      }
    `;

    const removeTodoResponse = await apiCall(removeTodoMutation, token);
    const { removeTodo } = removeTodoResponse.data;

    expect(removeTodo).toBeTruthy();

    const afterRemoveTodosResponse = await apiCall(query, token);
    const afterRemoveTodos = afterRemoveTodosResponse.data.todos;

    expect(afterRemoveTodos.length).toBe(initialTodos.length - 1);
  });

  afterAll(() => {
    token = null;
  });
});
