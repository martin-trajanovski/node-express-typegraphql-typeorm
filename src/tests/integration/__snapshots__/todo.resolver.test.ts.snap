// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test todos CRUD When we try to read the all todos with ADMIN role, it should match the snapshot 1`] = `
Array [
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Martin",
      "id": "a46bec40-a25c-4c1a-b976-2c72c5cf9a5b",
      "roles": Array [
        "ADMIN",
      ],
    },
    "id": "a12457a7-4d07-40f5-90a4-861954f122de",
    "title": "Test todo 1",
  },
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Martin",
      "id": "a46bec40-a25c-4c1a-b976-2c72c5cf9a5b",
      "roles": Array [
        "ADMIN",
      ],
    },
    "id": "57c60882-6696-4eae-b391-a44615615e6c",
    "title": "Test todo 2",
  },
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Test",
      "id": "2d17d0f0-68eb-40b2-9b30-0904048562ed",
      "roles": Array [
        "USER",
      ],
    },
    "id": "22de9fbd-f2ed-46d7-8a21-9f1f5a231a80",
    "title": "Test todo 3",
  },
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Test",
      "id": "2d17d0f0-68eb-40b2-9b30-0904048562ed",
      "roles": Array [
        "USER",
      ],
    },
    "id": "bdff5622-82d8-4e79-90f2-8470ee5fcf2f",
    "title": "Test todo 4",
  },
]
`;

exports[`Test todos CRUD When we try to read the todos, it should get all todos for specific user in deault order 1`] = `
Array [
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Test",
      "id": "2d17d0f0-68eb-40b2-9b30-0904048562ed",
      "roles": Array [
        "USER",
      ],
    },
    "id": "22de9fbd-f2ed-46d7-8a21-9f1f5a231a80",
    "title": "Test todo 3",
  },
  Object {
    "completed": false,
    "createdBy": Object {
      "firstName": "Test",
      "id": "2d17d0f0-68eb-40b2-9b30-0904048562ed",
      "roles": Array [
        "USER",
      ],
    },
    "id": "bdff5622-82d8-4e79-90f2-8470ee5fcf2f",
    "title": "Test todo 4",
  },
]
`;
